rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,212 @@
+}
+
+persistableStateService.register(MY_STATE_ID, { migrate, inject, extract });
+```
+
+```ts
+  const stateReadyForPersisting = persistableStateService.get(MY_STATE_ID),beforeSave(myState);
+  const stateReadyForConsuming = persistableStateService.get(MY_STATE_ID).afterLoad(stateReadyForPersisting);
+```
+
+# Motivation
+
+We need a formal way for authors of registry implementations to add migrations for this data. Since they may not own the saved object that their implementation ends up in, we can't rely on saved object migrations as it exists today (not to mention data that ends up in URLs that may need to be migrated).
+
+We also need to make sure that all the persited state containing references to saved objects extracts those before being saved and injects them later.
+
+# Detailed design
+
+```ts
+interface PersistableState extends Serializable {}
+
+interface PersistableStateDefinition {
+  id: string,
+  // migrate function receives state and version string and should return latest state version
+  // default is identity function
+  migrate: (state: PersistableState, version: string) => PersistableState,
+  // inject function receives state and a list of references and should return state with references injected
+  // default is identity function
+  inject: (state: PersistableState, references: SavedObjectReference[]) => PersistableState,
+  // extract function receives state and should return state with references extracted and array of references
+  // default returns same state with empty reference array
+  extract: (state: PersistableState) => [PersistableState, SavedObjectReference[]]
+}
+
+class PersistableStatePlugin {
+  setup(core, plugins) {
+    return {
+      // if one of the functions is not provided default is used 
+      // so we can always assume that all of the functions are available for every persistable state 
+      register: (id: string, definition: Partial<PersistableStateDefinition>) => void
+    }
+  }
+  start(core, plugins) {
+    return {
+      // get will always return. If requested persistable state is not defined it will return default set of functions.
+      get: (id: string) => PersistableStateDefinition,
+      // takes the state, extracts the references and returns them + version string
+      beforeSave: (id: string, state: PersistableState) => [PersistableState, SavedObjectReference[], string],
+      // takes the state, references and version and returns latest state version with references injected
+      afterLoad: (id: string, state: PersistableState, references: SavedObjectReference[], version: string) => PersistableState,
+    }
+  }
+}
+```
+
+
+# Consuming and edge cases
+
+When plugin A wants to store some state that it does not own it should check with persistableStateRegistry for a registered persistable state definition and execute `beforeSave` function.
+
+```ts
+  const expressionReadyForSaving = persistableStatePlugin.beforeSave('expression', expressionString);
```

Comment:
If the example is going to give a known id like this, I think we should encourage direct access via expression plugin. Otherwise, change it to use a variable?

```ts
const newState = persistableStatePlugin.beforeSave(id, oldState);
```